### PR TITLE
Scan the QR code of App to install app in team - Nodejs

### PR DESCRIPTION
**Pull request by NikitaJain-MSFT**
Generate the QR code of App to Install the application from mobile by scanning QR code.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,6 +64,7 @@ Here are all the samples which are deployed on Microsoft Azure and you can try i
 |9| Universal bots  | Teams catering bot demonstrating how to use Universal bots in Teams | [View](samples/bot-teams-catering/csharp) |  | |
 |10| Sequential workflow adaptive cards | Demonstrating on how to implement sequential flow, user specific view and upto date adaptive cards in bot. | [View][sequential#workflow#csharp] | [View][sequential#workflow#js]
 |11| Channel messages with RSC permissions | Demonstrating on how a bot can receive all channel messages with RSC without @mention. | [View][messageswithrsc#csharp] |[View][messageswithrsc#js]
+|13| QR App Installtion |This sample demos app installation using QR code which has applications app id |  | [View][qrappinstallation#nodejs]
```
Comment (v-chetsh) [inline README.md:67]: I think It should be 12th

File: README.md [hunk 2]
```diff
@@ -217,6 +218,7 @@ Here are all the samples which are deployed on Microsoft Azure and you can try i
 [messageswithrsc#csharp]:samples/bot-receive-channel-messages-withRSC/csharp
```
Comment (v-chetsh) [inline README.md:218]: verify if root md file is properly link with your project readme file.

File: samples/QR-app-installation/nodejs/appPackage/manifest.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  "id": "{{MicrosoftAppId}}",
+  "packageName": "com.microsoft.teams.qrappinstallation",
+  "developer": {
+    "name": "Microsoft",
+    "websiteUrl": "https://www.microsoft.com",
+    "privacyUrl": "https://www.microsoft.com/privacy",
+    "termsOfUseUrl": "https://www.microsoft.com/termsofuse"
+  },
+  "name": {
+    "short": "QRAppInstallation",
```
Comment (v-chetsh) [inline samples/QR-app-installation/nodejs/appPackage/manifest.json:14]: Add proper app name.

File: samples/QR-app-installation/nodejs/appPackage/manifest.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      ],
+      "isNotificationOnly": false
+    }
+  ],
+  "permissions": [ "identity", "messageTeamMembers" ],
+  "devicePermissions": [
+    "media"
+  ],
+  "validDomains": [
+    "*.ngrok.io",
```
Comment (v-chetsh) [inline samples/QR-app-installation/nodejs/appPackage/manifest.json:42]: In this sample you have mentioned .ngrok as valid domains, please check in C# example where you have added dummy data.

File: samples/QR-app-installation/nodejs/ssoOauthHelpler.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+        if (activity.type !== ActivityTypes.Invoke || activity.name !== tokenExchangeOperationName) {
+            throw new Error('TokenExchangeState can only be used with Invokes of signin/tokenExchange.');
+        }
+        const value = activity.value;
+
+        if (!value || !value.id) {
+            throw new Error('Invalid signin/tokenExchange. Missing activity.value.id.');
+        }
+        return `${channelId}/${conversationId}/${value.id}`;
```
Comment (v-chetsh) [inline samples/QR-app-installation/nodejs/ssoOauthHelpler.js:121]: Before every return statement make sure to add empty line.

File: samples/QR-app-installation/nodejs/views/installApp.ejs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                        else if (decodedText == "81fef3a6-72aa-4648-a763-de824aeafb7d") {
+                            appName = "Forms"
+                        }
+                        if (decodedText == "7299542a-1697-4ec1-812b-6b70065c0795") {
+                            appName = "Azure Boards"
+                        }
+                        if (decodedText == "0c5cfdbb-596f-4d39-b557-5d9516c94107") {
+                            appName = "App Studio"
+                        }
+                        $("#scannedText").text("Please click on confirm to install app -" + appName + "");
```
Comment (v-chetsh) [inline samples/QR-app-installation/nodejs/views/installApp.ejs:48]: one line space before it.

File: samples/QR-app-installation/nodejs/views/installApp.ejs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                        }
+                        else if (decodedText == "81fef3a6-72aa-4648-a763-de824aeafb7d") {
+                            appName = "Forms"
+                        }
+                        if (decodedText == "7299542a-1697-4ec1-812b-6b70065c0795") {
+                            appName = "Azure Boards"
+                        }
+                        if (decodedText == "0c5cfdbb-596f-4d39-b557-5d9516c94107") {
+                            appName = "App Studio"
+                        }
```
Comment (v-chetsh) [inline samples/QR-app-installation/nodejs/views/installApp.ejs:47]: Line space before every if statement.

File: samples/QR-app-installation/nodejs/ssoOauthHelpler.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        if (!tokenExchangeResponse || !tokenExchangeResponse.token) {
+            // The token could not be exchanged (which could be due to a consent requirement)
+            // Notify the sender that PreconditionFailed so they can respond accordingly.
+            await turnContext.sendActivity(
+                {
+                    type: ActivityTypes.InvokeResponse,
+                    value:
+                    {
+                        status: StatusCodes.PRECONDITION_FAILED,
+                        // TokenExchangeInvokeResponse
```
Comment (v-chetsh) [inline samples/QR-app-installation/nodejs/ssoOauthHelpler.js:83]: Remove if not required or add proper summary.

**Review by v-chetsh — CHANGES_REQUESTED** [review]
Reviewed and provided my feedback.

**Review by v-chetsh — APPROVED** [review]
Changes looks fine, approving it.

File: README.md [hunk 9]
```diff
@@ -218,6 +219,7 @@ Here are all the samples which are deployed on Microsoft Azure and you can try i
...
 [messageswithrsc#js]:samples/bot-receive-channel-messages-withRSC/nodejs
 [app-in-meeting#cs]:samples/app-in-meeting/csharp
+[qrappinstallation#nodejs]:samples/QR-app-installation/nodejs
```
Comment (Wajeed-msft) [inline README.md:222]: Please make the same changes (C# PR) for folder name and image in readme file.

**Review by Wajeed-msft — APPROVED** [review]

File: README.md (hunk 9 above)
Comment (AbhijeetThakur-msft) [reply README.md]: done

**Pull request merged**